### PR TITLE
Object cache named region no-op implementation

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Requires #1690 to be merged first.

The first 11 commits are from that PR, only the final two are new.

This sets up the named region async job interface in a no-op fashion.
The async interface itself is exercised while the async thread that is supposed to do loading and storing of these objects is a minimum no-op implementation.

Next PR will be for having the async thread actually loading and storing these object files.

### CODE REVIEW

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h [hunk 1]
```diff
@@ -128,17 +235,102 @@ namespace FEXCore::CodeSerialize {
...
+      CodeObjectSerializationConfig const &GetDefaultSerializationConfig() const {
+        return DefaultSerializationConfig;
+      }
+
+    protected:
+      friend class AsyncJobHandler;
+
+      // Return a default code header based off the default serialization config
+      CodeObjectSerializationHeader DefaultCodeHeader(uint64_t Base, uint64_t Offset) const {
+        CodeObjectSerializationHeader Header{};
```
Comment (neobrain) [inline External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:270]: This seems like a good place to use designated initializers. That way, it's harder to forget updating this code when adding new members in future changes, since the compiler can warn about it (e.g. gcc does this with `-Wextra`).
```suggestion
CodeObjectSerializationHeader Header{
        .Config = DefaultSerializationConfig,
        .OriginalBase = Base;
        .OriginalOffset = Offset;
        .NumCodeEntries = 0;
        .NumRelocationsTo = 0;
        .TotalRelocationsCount = 0;
};
```

File: External/FEXCore/Source/Interface/Core/ObjectCache/JobHandling.cpp [hunk 2]
```diff
@@ -13,11 +13,112 @@
...
+      // Lock the job ref counter so we can block anything attempting to use the entry before it is loaded
+      Entry->NamedJobRefCountMutex.lock();
+
+      CodeRegionMapType::iterator EntryIterator;
+      {
+        std::unique_lock lk {CodeObjectCacheService->GetEntryMapMutex()};
+
+        auto &EntryMap = CodeObjectCacheService->GetEntryMap();
+
+        auto it = EntryMap.emplace(Base, std::move(Entry));
```
Comment (neobrain) [inline External/FEXCore/Source/Interface/Core/ObjectCache/JobHandling.cpp:40]: The `!it.second` case can be slightly sped up by changing this to `EntryMap.emplace(Base, nullptr)`. The heap-allocating `std::make_unique<CodeRegionEntry>` only needs to be done in the `it.second == true` case then, which would explicitly overwrite the dummy entry created. (This assumes we can bear holding the lock a little bit longer in that case).

**Comment by skmp** [comment]
This will have to be modified for #1558 as it replaces the NameRegion interface, though that should be fairly easy to do

File: External/FEXCore/Source/Interface/Core/ObjectCache/JobHandling.cpp (hunk 2 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/JobHandling.cpp:40]: I'm not sure I'm following what you mean here. We're not inserting a dummy entry. We're inserting an entry with the minimal amount of data required to kick off the async job.

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h (hunk 1 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h]: Resolved to use designated initializers.

File: External/FEXCore/Source/Interface/Context/Context.h [hunk 3]
```diff
@@ -104,6 +104,7 @@ namespace FEXCore::Context {
       FEX_CONFIG_OPT(BlockJITNaming, BLOCKJITNAMING);
       FEX_CONFIG_OPT(ParanoidTSO, PARANOIDTSO);
       FEX_CONFIG_OPT(CacheObjectCodeCompilation, CACHEOBJECTCODECOMPILATION);
+      FEX_CONFIG_OPT(x87ReducedPrecision, X87REDUCEDPRECISION);
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Context/Context.h:107]: This seems unrelated?

Comment (skmp) [reply External/FEXCore/Source/Interface/Context/Context.h:107]: ah used for context format. We should also add to AOTIR.

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Context/Context.h:107]: It is necessary to have for `CodeObjectSerializationConfig` which pulls all of its configuration features from the Context Config location. Anything that effects codegen needs to be generated in that hash object.

File: External/FEXCore/Source/Interface/Core/ObjectCache/NamedRegionObjectHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace FEXCore::CodeSerialize {
+  NamedRegionObjectHandler::NamedRegionObjectHandler(FEXCore::Context::Context *ctx) {
+    DefaultSerializationConfig.Cookie = CODE_COOKIE;
+
+    // Initialize the Arch from CPUID
+    uint32_t Arch = ctx->CPUID.RunFunction(0x4000'0001, 0).eax & 0xF;
+    DefaultSerializationConfig.Arch = Arch;
+
+    DefaultSerializationConfig.MaxInstPerBlock = ctx->Config.MaxInstPerBlock;
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ObjectCache/NamedRegionObjectHandler.cpp:14]: `MaxInstPerBlock` and `Multiblock` also need to be included here?

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/NamedRegionObjectHandler.cpp:14]: Mixing them up should be fine, though we may or may not want that

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/NamedRegionObjectHandler.cpp:14]: Only one missing here is Multiblock since Multiblock isn't supported with the code cache yet.

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/NamedRegionObjectHandler.cpp:14]: Mutliblock is actually included in this PR, though neither `MaxInstPerBlock` nor `Multiblock` should change our ABI.

`HostFeatures` makes sense.

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h [hunk 5]
```diff
@@ -128,17 +234,101 @@ namespace FEXCore::CodeSerialize {
...
+        ));
+        ++NamedWorkQueueJobs;
+      }
+
+    private:
+      // Code version. If the code emission changes then this needs to increment
+      constexpr static uint32_t CODE_VERSION = 0x0;
+
+      // Default cookie header for the file header
+      constexpr static uint64_t CODE_COOKIE = FEXCore::IR::COOKIE_VERSION("FEXC", CODE_VERSION);
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:310]: Wouldn't "FEXO" be more consistent with IR's "FEXI" ?

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h [hunk 6]
```diff
@@ -128,17 +234,101 @@ namespace FEXCore::CodeSerialize {
...
+          Base,
+          Size,
+          std::move(Entry)
+        ));
+        ++NamedWorkQueueJobs;
+      }
+
+    private:
+      // Code version. If the code emission changes then this needs to increment
+      constexpr static uint32_t CODE_VERSION = 0x0;
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:307]: It's gonna be tricky to keep this always updated, as minor codegen changes are common. How about using the git hash?

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h (hunk 5 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:310]: Probably. It's just a cookie though, doesn't matter.

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h [hunk 7]
```diff
@@ -128,17 +234,101 @@ namespace FEXCore::CodeSerialize {
...
+          .TotalRelocationsCount = 0,
+        };
+      }
+
+      /**
+       * @brief Adds an asynchronous add named region work item to the object queue
+       *
+       * This adds the job that will do the loading of file resources and data tracking.
+       */
+      void AsyncAddNamedRegionWorkItem(const std::string &base, const std::string &filename, bool executable, CodeRegionMapType::iterator entry) {
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: Why do we need async versions of this?

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: Because we can't block the JIT on IO.

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h (hunk 6 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:307]: I was planning on release tag + incremental index but a git hash works just as well.

File: External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h (hunk 7 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: sure JIT->MMAP->Named Region.
It's significant IO. Add it to an async job queue to stop blocking JIT execution.

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: I think we should design the system in a way that substantial IO doesn't need to happen on file load.

Performance wise it is probably better, and far easier to reason about, small stutters on code file mmaps vs other complications or unpredictable performance later on.

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: The act of loading the index required to know what is in the file is enough IO to make the JIT slow down.
It's unfeasible to lower IO to the point that it doesn't impact JIT performance.

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: Can't the index be flatted down and then just mmap'd, similar to what AOTIR does?

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: Also WRT issue, it's probably best we keep separate "index" and "object code"/"ir code" files, as the indexes need to be rebuilt while the code files can just be appended to.

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: The entire file is just mmap'd already. flattening won't work since this file is multi-process shared and these caches get incrementally appended from multiple processes.

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: Flattening can work just fine with multi process, why wouldn't it?

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: (can just use a red/black tree that uses local offsets instead of pointers)

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h:284]: Well. It's not how I'm implementing it regardless.

File: External/FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h [hunk 8]
```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <cstdint>
+
+namespace FEXCore::CodeSerialize {
+  // If any of the config options mismatch on load then the cache won't be used
+  // Any of these will result in codegen changes
+  struct CodeObjectSerializationConfig {
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h:8]: Would be great to have this harmonized with the AOTIR format.

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h:8]: Could probably share some of the config but not all of it. Once this config pulls in host features then there is a mismatch between what config effects IR versus which config options effects assembly.

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h:8]: Yeah, we could split into two parts one for "IR-abi" and one for "object-abi".

Would also be nice if we used the same postfix notations in the cache files in both.

There's a further slight complication there with #1558, `MMAN` and `MTRACK` smc modes don't change the ABI at all.

Comment (skmp) [reply External/FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h:8]: (But are different `SMCChecks` values)

Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ObjectCache/CodeObjectSerializationConfig.h:8]: That's fine, doesn't need to be flawless.
